File: plugins/claude-swarm/agent-sdk/python/agents.py
"""
Claude Swarm - Agent Implementations
Implementacoes dos agentes usando Claude Agent SDK.
"""

import asyncio
from abc import ABC, abstractmethod
from typing import Optional, Dict, Any, List
from dataclasses import dataclass

from anthropic import Anthropic

from config import config, agent_config
from message_broker import broker, Message, MessageType


@dataclass
class AgentContext:
    """Contexto de execucao do agente."""
    task_id: str
    instruction: str
    metadata: Dict[str, Any]
    shared_state: Dict[str, Any]


class BaseAgent(ABC):
    """
    Classe base para todos os agentes do Swarm.
    """

    def __init__(self, agent_type: str):
        self.agent_type = agent_type
        self.agent_id = f"worker-{agent_type}"
        self.config = agent_config.get_config(agent_type)
        self.client = Anthropic(api_key=config.anthropic_api_key)
        self._running = False
        self._current_task: Optional[str] = None

    @property
    @abstractmethod
    def system_prompt(self) -> str:
        """System prompt especifico do agente."""
        pass

    @property
    def tools(self) -> List[Dict[str, Any]]:
        """Ferramentas disponiveis para o agente."""
        return self._get_tool_definitions()

    def _get_tool_definitions(self) -> List[Dict[str, Any]]:
        """Retorna definicoes de ferramentas baseado no tipo de agente."""
        all_tools = {
            "swarm_state_get": {
                "name": "swarm_state_get",
                "description": "Recupera estado compartilhado do Swarm",
                "input_schema": {
                    "type": "object",
                    "properties": {
                        "key": {"type": "string", "description": "Chave do estado"}
                    },
                    "required": ["key"]
                }
            },
            "swarm_store_result": {
                "name": "swarm_store_result",
                "description": "Armazena resultado da tarefa para o orchestrator",
                "input_schema": {
                    "type": "object",
                    "properties": {
                        "result": {"type": "string", "description": "Resultado em JSON"},
                        "status": {"type": "string", "enum": ["success", "partial", "failed"]}
                    },
                    "required": ["result"]
                }
            },
            "read_file": {
                "name": "read_file",
                "description": "Le conteudo de um arquivo",
                "input_schema": {
                    "type": "object",
                    "properties": {
                        "path": {"type": "string", "description": "Caminho do arquivo"}
                    },
                    "required": ["path"]
                }
            },
            "write_file": {
                "name": "write_file",
                "description": "Escreve conteudo em um arquivo",
                "input_schema": {
                    "type": "object",
                    "properties": {
                        "path": {"type": "string", "description": "Caminho do arquivo"},
                        "content": {"type": "string", "description": "Conteudo"}
                    },
                    "required": ["path", "content"]
                }
            },
            "search_code": {
                "name": "search_code",
                "description": "Busca padrao em arquivos de codigo",
                "input_schema": {
                    "type": "object",
                    "properties": {
                        "pattern": {"type": "string", "description": "Padrao regex"},
                        "path": {"type": "string", "description": "Diretorio base"}
                    },
                    "required": ["pattern"]
                }
            },
            "run_command": {
                "name": "run_command",
                "description": "Executa comando shell",
                "input_schema": {
                    "type": "object",
                    "properties": {
                        "command": {"type": "string", "description": "Comando a executar"}
                    },
                    "required": ["command"]
                }
            }
        }

        allowed_tools = self.config.get("tools", [])
        return [all_tools[t] for t in allowed_tools if t in all_tools]

    async def _handle_tool_call(
        self,
        tool_name: str,
        tool_input: Dict[str, Any],
        context: AgentContext
    ) -> str:
        """Processa chamada de ferramenta."""

        if tool_name == "swarm_state_get":
            result = await broker.get_state(tool_input["key"])
            return str(result) if result else "null"

        elif tool_name == "swarm_store_result":
            import json
            result = json.loads(tool_input["result"])
            status = tool_input.get("status", "success")
            await broker.store_result(context.task_id, result, status)
            return f"Resultado armazenado com status: {status}"

        elif tool_name == "read_file":
            try:
                with open(tool_input["path"], "r") as f:
                    return f.read()
            except Exception as e:
                return f"Erro: {e}"

        elif tool_name == "write_file":
            try:
                with open(tool_input["path"], "w") as f:
                    f.write(tool_input["content"])
                return f"Arquivo salvo: {tool_input['path']}"
            except Exception as e:
                return f"Erro: {e}"

        elif tool_name == "search_code":
            import subprocess
            try:
                result = subprocess.run(
                    ["grep", "-r", tool_input["pattern"], tool_input.get("path", ".")],
                    capture_output=True,
                    text=True,
                    timeout=30
                )
                return result.stdout or "Nenhum resultado"
            except Exception as e:
                return f"Erro: {e}"

        elif tool_name == "run_command":
            import subprocess
            try:
                result = subprocess.run(
                    tool_input["command"],
                    shell=True,
                    capture_output=True,
                    text=True,
                    timeout=60
                )
                return f"stdout: {result.stdout}\nstderr: {result.stderr}"
            except Exception as e:
                return f"Erro: {e}"

        return f"Ferramenta desconhecida: {tool_name}"

    async def process_task(self, context: AgentContext) -> Dict[str, Any]:
        """
        Processa uma tarefa usando o Claude.

        Args:
            context: Contexto da tarefa

        Returns:
            Resultado do processamento
        """
        self._current_task = context.task_id
        messages = [{"role": "user", "content": context.instruction}]

        try:
            # Loop de conversacao com ferramentas
            while True:
                response = self.client.messages.create(
                    model=self.config["model"],
                    max_tokens=self.config["max_tokens"],
                    system=self.system_prompt,
                    tools=self.tools if self.tools else None,
                    messages=messages
                )

                # Verificar se terminou
                if response.stop_reason == "end_turn":
                    # Extrair texto final
                    text_blocks = [b.text for b in response.content if hasattr(b, 'text')]
                    return {
                        "success": True,
                        "output": "\n".join(text_blocks),
                        "model": self.config["model"]
                    }

                # Processar tool calls
                if response.stop_reason == "tool_use":
                    messages.append({"role": "assistant", "content": response.content})

                    tool_results = []
                    for block in response.content:
                        if block.type == "tool_use":
                            result = await self._handle_tool_call(
                                block.name,
                                block.input,
                                context
                            )
                            tool_results.append({
                                "type": "tool_result",
                                "tool_use_id": block.id,
                                "content": result
                            })

                    messages.append({"role": "user", "content": tool_results})
                else:
                    # Outros stop reasons
                    break

        except Exception as e:
            return {
                "success": False,
                "error": str(e),
                "model": self.config["model"]
            }
        finally:
            self._current_task = None

    async def handle_message(self, message: Message) -> None:
        """
        Handler para mensagens recebidas.

        Args:
            message: Mensagem recebida
        """
        if message.type == MessageType.TASK:
            print(f"[{self.agent_id}] Recebida tarefa: {message.id}")

            context = AgentContext(
                task_id=message.id,
                instruction=message.payload.get("instruction", ""),
                metadata=message.metadata,
                shared_state={}
            )

            result = await self.process_task(context)
            await broker.store_result(message.id, result)

            print(f"[{self.agent_id}] Tarefa {message.id} completada")

        elif message.type == MessageType.BROADCAST:
            action = message.payload.get("action", "")
            print(f"[{self.agent_id}] Broadcast recebido: {action}")

            if action == "shutdown":
                self._running = False
            elif action == "status":
                status = {
                    "agent_id": self.agent_id,
                    "type": self.agent_type,
                    "running": self._running,
                    "current_task": self._current_task
                }
                await broker.set_state(f"status:{self.agent_id}", status, ttl_seconds=60)

    async def start(self) -> None:
        """Inicia o agente."""
        self._running = True
        await broker.connect()

        # Registrar
        await broker.set_state(f"agents:{self.agent_id}", {
            "type": self.agent_type,
            "status": "running",
            "started_at": asyncio.get_event_loop().time()
        })

        # Inscrever em canais
        await broker.subscribe(
            ["tasks", "broadcast"],
            self.handle_message
        )

        # Iniciar heartbeat
        asyncio.create_task(self._heartbeat_loop())

        print(f"[{self.agent_id}] Agente iniciado")

        # Loop principal
        await broker.listen()

    async def _heartbeat_loop(self) -> None:
        """Loop de heartbeat."""
        while self._running:
            await broker.heartbeat()
            await asyncio.sleep(config.heartbeat_interval)

    async def stop(self) -> None:
        """Para o agente."""
        self._running = False
        await broker.set_state(f"agents:{self.agent_id}", {
            "type": self.agent_type,
            "status": "stopped"
        })
        await broker.disconnect()
        print(f"[{self.agent_id}] Agente parado")


# =============================================================================
# IMPLEMENTACOES ESPECIFICAS
# =============================================================================

class AnalystAgent(BaseAgent):
    """Agente especializado em analise."""

    def __init__(self):
        super().__init__("analyst")

    @property
    def system_prompt(self) -> str:
        return """Voce e um Analista especializado no Claude Swarm.

Sua especialidade:
- Analisar codigo, arquitetura e requisitos
- Identificar padroes, problemas e oportunidades
- Documentar descobertas de forma clara

Diretrizes:
- Seja objetivo e baseado em evidencias
- Forneca analises estruturadas
- Destaque riscos e recomendacoes

Voce trabalha como parte de um swarm de agentes. Foque na sua especialidade
e contribua com insights de analise para a tarefa coletiva."""


class CoderAgent(BaseAgent):
    """Agente especializado em codificacao."""

    def __init__(self):
        super().__init__("coder")

    @property
    def system_prompt(self) -> str:
        return """Voce e um Desenvolvedor especializado no Claude Swarm.

Sua especialidade:
- Escrever codigo limpo e eficiente
- Implementar features e corrigir bugs
- Seguir padroes e boas praticas

Diretrizes:
- Codigo deve ser legivel e bem documentado
- Siga os padroes do projeto existente
- Considere performance e seguranca

Voce trabalha como parte de um swarm de agentes. Foque na implementacao
e contribua com codigo de qualidade para a tarefa coletiva."""


class ReviewerAgent(BaseAgent):
    """Agente especializado em revisao."""

    def __init__(self):
        super().__init__("reviewer")

    @property
    def system_prompt(self) -> str:
        return """Voce e um Revisor especializado no Claude Swarm.

Sua especialidade:
- Revisar codigo para qualidade e seguranca
- Identificar bugs, vulnerabilidades e melhorias
- Sugerir refatoracoes e otimizacoes

Diretrizes:
- Seja construtivo e especifico
- Priorize issues por severidade
- Forneca sugestoes concretas de melhoria

Voce trabalha como parte de um swarm de agentes. Foque na revisao
e contribua com feedback de qualidade para a tarefa coletiva."""


class TesterAgent(BaseAgent):
    """Agente especializado em testes."""

    def __init__(self):
        super().__init__("tester")

    @property
    def system_prompt(self) -> str:
        return """Voce e um Engenheiro de Testes especializado no Claude Swarm.

Sua especialidade:
- Criar e executar testes automatizados
- Validar requisitos e comportamentos
- Identificar casos de borda e falhas

Diretrizes:
- Cobertura completa de casos de uso
- Testes devem ser reproduziveis
- Documente cenarios e resultados

Voce trabalha como parte de um swarm de agentes. Foque em testes
e contribua com validacao de qualidade para a tarefa coletiva."""


class ResearcherAgent(BaseAgent):
    """Agente especializado em pesquisa."""

    def __init__(self):
        super().__init__("researcher")

    @property
    def system_prompt(self) -> str:
        return """Voce e um Pesquisador especializado no Claude Swarm.

Sua especialidade:
- Pesquisar informacoes e documentacao
- Coletar dados e referencias
- Sintetizar conhecimento relevante

Diretrizes:
- Busque fontes confiaveis
- Valide informacoes encontradas
- Apresente descobertas de forma estruturada

Voce trabalha como parte de um swarm de agentes. Foque na pesquisa
e contribua com conhecimento relevante para a tarefa coletiva."""


# Factory
AGENT_CLASSES = {
    "analyst": AnalystAgent,
    "coder": CoderAgent,
    "reviewer": ReviewerAgent,
    "tester": TesterAgent,
    "researcher": ResearcherAgent
}


def create_agent(agent_type: str) -> BaseAgent:
    """
    Cria instancia de agente pelo tipo.

    Args:
        agent_type: Tipo do agente

    Returns:
        Instancia do agente
    """
    if agent_type not in AGENT_CLASSES:
        raise ValueError(f"Tipo de agente desconhecido: {agent_type}")
    return AGENT_CLASSES[agent_type]()

File: plugins/claude-swarm/agent-sdk/typescript/src/agents.ts
/**
 * Claude Swarm - Agent Implementations
 * Implementacoes dos agentes usando Claude Agent SDK.
 */

import Anthropic from '@anthropic-ai/sdk';
import { config, getAgentConfig, AgentTypeConfig } from './config';
import { broker } from './message-broker';
import { Message, MessageType, AgentContext, ToolDefinition } from './types';
import { execSync } from 'child_process';
import * as fs from 'fs';

export abstract class BaseAgent {
  protected agentType: string;
  protected agentId: string;
  protected config: AgentTypeConfig;
  protected client: Anthropic;
  protected running = false;
  protected currentTask: string | null = null;

  constructor(agentType: string) {
    this.agentType = agentType;
    this.agentId = `worker-${agentType}`;
    this.config = getAgentConfig(agentType);
    this.client = new Anthropic({ apiKey: config.anthropicApiKey });
  }

  abstract get systemPrompt(): string;

  get tools(): ToolDefinition[] {
    return this.getToolDefinitions();
  }

  private getToolDefinitions(): ToolDefinition[] {
    const allTools: Record<string, ToolDefinition> = {
      swarm_state_get: {
        name: 'swarm_state_get',
        description: 'Recupera estado compartilhado do Swarm',
        input_schema: {
          type: 'object',
          properties: {
            key: { type: 'string', description: 'Chave do estado' },
          },
          required: ['key'],
        },
      },
      swarm_store_result: {
        name: 'swarm_store_result',
        description: 'Armazena resultado da tarefa para o orchestrator',
        input_schema: {
          type: 'object',
          properties: {
            result: { type: 'string', description: 'Resultado em JSON' },
            status: { type: 'string', enum: ['success', 'partial', 'failed'] },
          },
          required: ['result'],
        },
      },
      read_file: {
        name: 'read_file',
        description: 'Le conteudo de um arquivo',
        input_schema: {
          type: 'object',
          properties: {
            path: { type: 'string', description: 'Caminho do arquivo' },
          },
          required: ['path'],
        },
      },
      write_file: {
        name: 'write_file',
        description: 'Escreve conteudo em um arquivo',
        input_schema: {
          type: 'object',
          properties: {
            path: { type: 'string', description: 'Caminho do arquivo' },
            content: { type: 'string', description: 'Conteudo' },
          },
          required: ['path', 'content'],
        },
      },
      search_code: {
        name: 'search_code',
        description: 'Busca padrao em arquivos de codigo',
        input_schema: {
          type: 'object',
          properties: {
            pattern: { type: 'string', description: 'Padrao regex' },
            path: { type: 'string', description: 'Diretorio base' },
          },
          required: ['pattern'],
        },
      },
      run_command: {
        name: 'run_command',
        description: 'Executa comando shell',
        input_schema: {
          type: 'object',
          properties: {
            command: { type: 'string', description: 'Comando a executar' },
          },
          required: ['command'],
        },
      },
    };

    const allowedTools = this.config.tools;
    return allowedTools
      .filter((t) => t in allTools)
      .map((t) => allTools[t]);
  }

  protected async handleToolCall(
    toolName: string,
    toolInput: Record<string, unknown>,
    context: AgentContext
  ): Promise<string> {
    switch (toolName) {
      case 'swarm_state_get': {
        const result = await broker.getState(toolInput.key as string);
        return result ? JSON.stringify(result) : 'null';
      }

      case 'swarm_store_result': {
        const result = JSON.parse(toolInput.result as string);
        const status = (toolInput.status as string) || 'success';
        await broker.storeResult(context.taskId, result, status);
        return `Resultado armazenado com status: ${status}`;
      }

      case 'read_file': {
        try {
          return fs.readFileSync(toolInput.path as string, 'utf-8');
        } catch (e) {
          return `Erro: ${e}`;
        }
      }

      case 'write_file': {
        try {
          fs.writeFileSync(toolInput.path as string, toolInput.content as string);
          return `Arquivo salvo: ${toolInput.path}`;
        } catch (e) {
          return `Erro: ${e}`;
        }
      }

      case 'search_code': {
        try {
          const result = execSync(
            `grep -r "${toolInput.pattern}" ${toolInput.path || '.'}`,
            { timeout: 30000, encoding: 'utf-8' }
          );
          return result || 'Nenhum resultado';
        } catch (e) {
          return `Erro: ${e}`;
        }
      }

      case 'run_command': {
        try {
          const result = execSync(toolInput.command as string, {
            timeout: 60000,
            encoding: 'utf-8',
          });
          return result;
        } catch (e) {
          return `Erro: ${e}`;
        }
      }

      default:
        return `Ferramenta desconhecida: ${toolName}`;
    }
  }

  async processTask(context: AgentContext): Promise<Record<string, unknown>> {
    this.currentTask = context.taskId;
    const messages: Anthropic.MessageParam[] = [
      { role: 'user', content: context.instruction },
    ];

    try {
      while (true) {
        const response = await this.client.messages.create({
          model: this.config.model,
          max_tokens: this.config.maxTokens,
          system: this.systemPrompt,
          tools: this.tools.length > 0 ? this.tools as Anthropic.Tool[] : undefined,
          messages,
        });

        if (response.stop_reason === 'end_turn') {
          const textBlocks = response.content
            .filter((b): b is Anthropic.TextBlock => b.type === 'text')
            .map((b) => b.text);
          return {
            success: true,
            output: textBlocks.join('\n'),
            model: this.config.model,
          };
        }

        if (response.stop_reason === 'tool_use') {
          messages.push({ role: 'assistant', content: response.content });

          const toolResults: Anthropic.ToolResultBlockParam[] = [];
          for (const block of response.content) {
            if (block.type === 'tool_use') {
              const result = await this.handleToolCall(
                block.name,
                block.input as Record<string, unknown>,
                context
              );
              toolResults.push({
                type: 'tool_result',
                tool_use_id: block.id,
                content: result,
              });
            }
          }

          messages.push({ role: 'user', content: toolResults });
        } else {
          break;
        }
      }

      return {
        success: false,
        error: 'Execucao interrompida inesperadamente',
        model: this.config.model,
      };
    } catch (e) {
      return {
        success: false,
        error: String(e),
        model: this.config.model,
      };
    } finally {
      this.currentTask = null;
    }
  }

  async handleMessage(message: Message): Promise<void> {
    if (message.type === MessageType.TASK) {
      console.log(`[${this.agentId}] Recebida tarefa: ${message.id}`);

      const context: AgentContext = {
        taskId: message.id,
        instruction: (message.payload.instruction as string) || '',
        metadata: message.metadata,
        sharedState: {},
      };

      const result = await this.processTask(context);
      await broker.storeResult(message.id, result);

      console.log(`[${this.agentId}] Tarefa ${message.id} completada`);
    } else if (message.type === MessageType.BROADCAST) {
      const action = message.payload.action as string;
      console.log(`[${this.agentId}] Broadcast recebido: ${action}`);

      if (action === 'shutdown') {
        this.running = false;
      } else if (action === 'status') {
        const status = {
          agentId: this.agentId,
          type: this.agentType,
          running: this.running,
          currentTask: this.currentTask,
        };
        await broker.setState(`status:${this.agentId}`, status, 60);
      }
    }
  }

  async start(): Promise<void> {
    this.running = true;
    await broker.connect();

    await broker.setState(`agents:${this.agentId}`, {
      type: this.agentType,
      status: 'running',
      startedAt: Date.now(),
    });

    await broker.subscribe(['tasks', 'broadcast'], (msg) => this.handleMessage(msg));

    this.heartbeatLoop();

    console.log(`[${this.agentId}] Agente iniciado`);

    await broker.listen();
  }

  private heartbeatLoop(): void {
    const interval = setInterval(async () => {
      if (!this.running) {
        clearInterval(interval);
        return;
      }
      await broker.heartbeat();
    }, config.heartbeatInterval * 1000);
  }

  async stop(): Promise<void> {
    this.running = false;
    await broker.setState(`agents:${this.agentId}`, {
      type: this.agentType,
      status: 'stopped',
    });
    await broker.disconnect();
    console.log(`[${this.agentId}] Agente parado`);
  }
}

// =============================================================================
// IMPLEMENTACOES ESPECIFICAS
// =============================================================================

export class AnalystAgent extends BaseAgent {
  constructor() {
    super('analyst');
  }

  get systemPrompt(): string {
    return `Voce e um Analista especializado no Claude Swarm.

Sua especialidade:
- Analisar codigo, arquitetura e requisitos
- Identificar padroes, problemas e oportunidades
- Documentar descobertas de forma clara

Diretrizes:
- Seja objetivo e baseado em evidencias
- Forneca analises estruturadas
- Destaque riscos e recomendacoes

Voce trabalha como parte de um swarm de agentes. Foque na sua especialidade
e contribua com insights de analise para a tarefa coletiva.`;
  }
}

export class CoderAgent extends BaseAgent {
  constructor() {
    super('coder');
  }

  get systemPrompt(): string {
    return `Voce e um Desenvolvedor especializado no Claude Swarm.

Sua especialidade:
- Escrever codigo limpo e eficiente
- Implementar features e corrigir bugs
- Seguir padroes e boas praticas

Diretrizes:
- Codigo deve ser legivel e bem documentado
- Siga os padroes do projeto existente
- Considere performance e seguranca

Voce trabalha como parte de um swarm de agentes. Foque na implementacao
e contribua com codigo de qualidade para a tarefa coletiva.`;
  }
}

export class ReviewerAgent extends BaseAgent {
  constructor() {
    super('reviewer');
  }

  get systemPrompt(): string {
    return `Voce e um Revisor especializado no Claude Swarm.

Sua especialidade:
- Revisar codigo para qualidade e seguranca
- Identificar bugs, vulnerabilidades e melhorias
- Sugerir refatoracoes e otimizacoes

Diretrizes:
- Seja construtivo e especifico
- Priorize issues por severidade
- Forneca sugestoes concretas de melhoria

Voce trabalha como parte de um swarm de agentes. Foque na revisao
e contribua com feedback de qualidade para a tarefa coletiva.`;
  }
}

export class TesterAgent extends BaseAgent {
  constructor() {
    super('tester');
  }

  get systemPrompt(): string {
    return `Voce e um Engenheiro de Testes especializado no Claude Swarm.

Sua especialidade:
- Criar e executar testes automatizados
- Validar requisitos e comportamentos
- Identificar casos de borda e falhas

Diretrizes:
- Cobertura completa de casos de uso
- Testes devem ser reproduziveis
- Documente cenarios e resultados

Voce trabalha como parte de um swarm de agentes. Foque em testes
e contribua com validacao de qualidade para a tarefa coletiva.`;
  }
}

export class ResearcherAgent extends BaseAgent {
  constructor() {
    super('researcher');
  }

  get systemPrompt(): string {
    return `Voce e um Pesquisador especializado no Claude Swarm.

Sua especialidade:
- Pesquisar informacoes e documentacao
- Coletar dados e referencias
- Sintetizar conhecimento relevante

Diretrizes:
- Busque fontes confiaveis
- Valide informacoes encontradas
- Apresente descobertas de forma estruturada

Voce trabalha como parte de um swarm de agentes. Foque na pesquisa
e contribua com conhecimento relevante para a tarefa coletiva.`;
  }
}

// Factory
const AGENT_CLASSES: Record<string, new () => BaseAgent> = {
  analyst: AnalystAgent,
  coder: CoderAgent,
  reviewer: ReviewerAgent,
  tester: TesterAgent,
  researcher: ResearcherAgent,
};

export function createAgent(agentType: string): BaseAgent {
  const AgentClass = AGENT_CLASSES[agentType];
  if (!AgentClass) {
    throw new Error(`Tipo de agente desconhecido: ${agentType}`);
  }
  return new AgentClass();
}
